### PR TITLE
Added more control to some weighted experiments

### PR DESCRIPTION
**Pull request by regislebrun**
It concerns the LHSExperiment and the LowDiscrepancyExperiment. More specifically:
+ The LHSExperiment can now generate a new cell shuffle for each call to generate()
+ The LHSExperiment can now select the probabilistic center (ie the median point) of each cell instead of a random point for each cell
+ The LowDiscrepancyExperiment can now keep the current state of the underlying low-discrepancy experiment when the user change the distribution instead of reinitializing the sequence. It is mandatory eg for sequential Monte Carlo algorithms for robust optimization.

### CODE REVIEW

File: python/src/LowDiscrepancyExperiment_doc.i.in [hunk 1]
```diff
@@ -56,3 +106,27 @@ Returns
...
     Sequence of points :math:`(u_1, \cdots, u_N)` with low discrepancy."
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::LowDiscrepancyExperiment::getInitialize
+"Return the value of the *initialize* flag.
+
+Returns
+-------
+bool : value of the *initialize* flag."
```
Comment (jschueller) [inline python/src/LowDiscrepancyExperiment_doc.i.in]: initialize : bool

File: lib/src/Uncertainty/Algorithm/WeightedExperiments/LowDiscrepancyExperiment.cxx [hunk 2]
```diff
@@ -94,8 +116,9 @@ void LowDiscrepancyExperiment::setDistribution(const Distribution & distribution
   marginals_ = DistributionCollection(dimension);
   // Get the marginal distributions
   for (UnsignedInteger i = 0; i < dimension; ++ i) marginals_[i] = distribution.getMarginal(i);
-  // initialize the low-discrepancy sequence
-  sequence_.initialize(dimension);
+  // initialize the low-discrepancy sequence if asked for or mandatory (dimension changed)
+  if (initialize_ || (dimension != getDistribution().getDimension()))
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/WeightedExperiments/LowDiscrepancyExperiment.cxx]: well done!

**Comment by jschueller** [comment]
+1

**Comment by jschueller** [comment]
maybe this is worth mentionning in the changelog ?

File: python/src/LHSExperiment_doc.i.in [hunk 3]
```diff
@@ -13,6 +13,16 @@ distribution : :class:`~openturns.Distribution`
     set of input data.
 size : positive int
     Number :math:`\mathrm{card}\,I` of points that will be generated in the experiment.
+initialize : bool
+    Flag to tell if the shuffle must be regenerated at each call to generate
+    or not.
+    Default is *False*: the shuffle is generated once and for all.
```
Comment (dbarbier) [inline python/src/LHSExperiment_doc.i.in]: I find this member name misleading (alwaysShuffle?); moreover the same member name is added to LowDiscrepancyExperiment, but with a different default value, this is very confusing.

File: python/src/LHSExperiment_doc.i.in [hunk 4]
```diff
@@ -13,6 +13,16 @@ distribution : :class:`~openturns.Distribution`
...
+
+randomShift : bool
+    Flag to tell if the point selected in each cell of the shuffle is the center
+    of the cell (*randomshift* is *True*) or if it is drawn wrt to the restriction
```
Comment (dbarbier) [inline python/src/LHSExperiment_doc.i.in]: Center of the cell corresponds to randomShift being False, not True

File: python/src/LowDiscrepancyExperiment_doc.i.in [hunk 5]
```diff
@@ -22,6 +22,11 @@ distribution :
     :math:`\Xi_i =F_i^{-1}(u_i)` to generate points :math:`(\Xi_i)_{i\in I}`
     according to the distribution :math:`\mu`. The weights are all equal to
     :math:`1/N`.
+initialize : bool
+    Flag to tell if the low discrepancy sequence must be restarted from
+    its initial state at each change of distribution or not.
+    Default is *True*: the sequence is restarted at each change of
+    distribution.
```
Comment (dbarbier) [inline python/src/LowDiscrepancyExperiment_doc.i.in]: Again, member name looks strange to me.  restart?

File: lib/src/Uncertainty/Algorithm/WeightedExperiments/LHSExperiment.cxx [hunk 6]
```diff
@@ -86,15 +111,14 @@ NumericalSample LHSExperiment::generate()
   shuffle_ = getShuffle();
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/WeightedExperiments/LHSExperiment.cxx]: @dbarbier Shouldn't it just make a copy ?

**Comment by regislebrun** [comment]
@dbarbier I didn't know this construction, much better than assigning shuffle_ to itself! I have the same kind of assignment in DistributionImplementation when I call getMean() only to insure that mean_ is up to date.
@jschueller I am not very fond of local copies when they are not needed. The member is here to be used after all!

**Comment by dbarbier** [comment]
+1

**Pull request merged**